### ISSUE TITLE
mkl performance

### ISSUE THREAD

**Issue by epsi1on** [wo80/csparse-interop#20]
Hi,
i was testing the mkl feature with sparse matrices, did got a very high performance with magnitude of ~25x against C# factorization.

Here is the matrix information:
Square matrix with dimension of ~23k and nonzero count of ~263k of real type (created from a finite element model).

[mtx.zip](https://github.com/user-attachments/files/18281461/mtx.zip)

|Solver|Ordering|Runtime (milisec)|
|-|-|-|
|Pardiso|`PardisoOrdering.NestedDissection`|599|
|Pardiso|`PardisoOrdering.MinimumDegree`|718|
|SparseCholesky in CSparse.NET|`ColumnOrdering.MinimumDegreeAtPlusA`|14779 (14.7 seconds)|

As the[ comparison in wiki](https://github.com/wo80/csparse-interop/wiki/Test-Results) shows ~4x magnitude in performance against C# code, I am thinking that maybe i am missing something in my code?

Thanks

---

Here is the pardiso config, taken from this file:

```
 int mtype = symmetric ? PardisoMatrixType.RealStructurallySymmetric : PardisoMatrixType.RealNonsymmetric;

 var solver = new Pardiso(matrix, mtype);

 var options = solver.Options;

 // Fill-in reordering from METIS.
 options.ColumnOrderingMethod = PardisoOrdering.NestedDissection;

 // Max numbers of iterative refinement steps.
 options.IterativeRefinement = 2;

 // Perturb the pivot elements with 1E-13.
 options.PivotingPerturbation = 13;

 // Use nonsymmetric permutation and scaling MPS.
 options.Scaling = true;

 // Maximum weighted matching algorithm is switched-on (default for non-symmetric).
 options.WeightedMatching = true;

 if (symmetric)
 {
     // Maximum weighted matching algorithm is switched-off (default for symmetric).
     // Try to enable in case of inappropriate accuracy.
     options.WeightedMatching = false;
 }

 options.ZeroBasedIndexing = true;
 options.CheckMatrix = true;

```

**Comment by wo80** [comment]
Here's what I get running the test program with your data on my Ryzen 3600 CPU:
```
Testing CSparse.NET Cholesky (symmetric) ... [fill-in: 61.7] [29.570s] OK
Testing CXSparse Cholesky (symmetric) ... [fill-in: 61.7] [16.448s] OK
Testing UMFPACK (symmetric) ... [fill-in: 123.2] [9.759s] OK
Testing CHOLMOD (symmetric) ... [fill-in: 35.1] [1.579s] OK
Testing SuperLU (symmetric) ... [fill-in: 117.9] [5.062s] OK
Testing PARDISO (symmetric) ... [fill-in: 67.9] [0.432s] OK
```
Comparing CXSparse to CSparse.NET (which both use the same algorithm/code), I'd say the performance difference is what you can expect for native and managed code.

For the other solvers it's hard to compare because they use different algorithms. And PARDISO and CHOLMOD are probably some of the most optimized direct solvers, so it's no surprise to me that the performance is a lot better.

> As the [comparison in wiki](https://github.com/wo80/csparse-interop/wiki/Test-Results) shows ~4x magnitude in performance against C# code, I am thinking that maybe i am missing something in my code?

Keep in mind that sparse direct solvers don't have linear time complexity, see for example https://scicomp.stackexchange.com/a/16322. So while for the wiki benchmark, using a 2500x2500 matrix, we might get a 4x performance difference, the difference for larger matrices will most likely be higher.

**Comment by epsi1on** [comment]
Kindly thanks for information.

**Comment by epsi1on** [comment]
I was thinking maybe the PARDISO is using iterative method or hybrid method for solving the equations behind the scene?
but CSparse is using direct method (factorization), and that is the reason MKL have ~30x higher performance.

**Comment by wo80** [comment]
> Since PARDISO is even faster than CholMOD on GPU, it is a little bit confusing...

SuiteSparse is an opensource project developed basically by one person. I don't have any insights into the MKL/PARDISO development process, but it's probably more than one dev, getting paid and with a lot more resources ;-)

> I was thinking maybe the PARDISO is using iterative method or hybrid method for solving the equations behind the scene?

I don't think the iterative solver of PARDISO is used here. The `PardisoContext` isn't configured to use it and you can always check the value `solver.Options.iparm[19]` to see the status (only while debugging, it's not public - maybe it should be public...). The value of that entry is `0`, meaning no CG/CGS iterations were performed. See

* https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2025-0/onemkl-pardiso-parallel-direct-sparse-solver-iface.html#id-d756309e953
* https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2025-0/pardiso-iparm-parameter.html#IPARM20

**Comment by wo80** [comment]
Btw, the performance of PARDISO gets even better if you make sure to pass only the lower part of your symmetric matrix and set `mtype = PardisoMatrixType.RealSymmetricPositiveDefinite`:
```
Testing CHOLMOD (symmetric) ... [fill-in: 35.1] [1.585s] OK
Testing PARDISO ... [fill-in: 67.9] [0.548s] OK
Testing PARDISO (symmetric) ... [fill-in: 34.9] [0.293s] OK
```

**Comment by epsi1on** [comment]
> SuiteSparse is an opensource project developed basically by one person. I don't have any insights into the MKL/PARDISO development process, but it's probably more than one dev, getting paid and with a lot more resources ;-)
It definitely makes sense.

> I don't think the iterative solver of PARDISO is used here. The `PardisoContext` isn't configured to use it and you can always check the value `solver.Options.iparm[19]` to see the status (only while debugging, it's not public - maybe it should be public...). The value of that entry is `0`, meaning no CG/CGS iterations were performed. See

Thanks for detailed information. 
Is it possible to solve for more than one right hand side at same time with MKL direct SPD solver? If it is not possible, then I think the feature of Direct method of MKL is useless to me (and many other users), CG method is better for SPD matrix then.

Cheers

**Comment by epsi1on** [comment]
> > SuiteSparse is an opensource project developed basically by one person. I don't have any insights into the MKL/PARDISO development process, but it's probably more than one dev, getting paid and with a lot more resources ;-)
> > It definitely makes sense.

It definitely makes sense.

> > I don't think the iterative solver of PARDISO is used here. The `PardisoContext` isn't configured to use it and you can always check the value `solver.Options.iparm[19]` to see the status (only while debugging, it's not public - maybe it should be public...). The value of that entry is `0`, meaning no CG/CGS iterations were performed. See
> 
> Thanks for detailed information. Is it possible to solve for more than one right hand side at same time with MKL direct SPD solver? If it is not possible, then I think the feature of Direct method of MKL is useless to me (and many other users), CG method is better for SPD matrix then.
> 
> Cheers

**Comment by wo80** [comment]
> Is it possible to solve for more than one right hand side at same time with MKL direct SPD solver?

Yes, it's not part of the `ISolver` interface, since not all solvers support it, but PARDISO does, see

https://github.com/wo80/csparse-interop/blob/4996c6a8ccddf36e59c9c0049b69ffa961a8a0eb/CSparse.Interop/Interop/MKL/Pardiso/PardisoContext.cs#L117-L125
